client/my-sites/checkout/checkout/billing-details-form.jsx

```diff
@@ -0,0 +1,379 @@
+	constructor( props ) {
+		super( props );
+		this.debounced = {
+			validateBillingDetails: debounce( this.validateBillingDetails, 300, { leading: true } ),
+			validateDomainDetails: debounce( this.validateDomainDetails, 300, { leading: true } ),
+		};
+		this.state = {
+			showDomainContactDetails: false,
+			billingDetails: null,
+		};
+	}
+
+	componentDidMount() {
+		if ( this.props.recordTracksEvent ) {
+			this.props.recordTracksEvent( 'calypso_checkout_billing_contact_information_view' );
+		}
+	}
+
+	validateBillingDetails = ( fieldValues, onComplete ) => {
+		const validationHandler = ( error, data ) => {
+			const messages = ( data && data.messages ) || {};
+
+			if ( ! error ) {
+				this.setState( { billingDetails: fieldValues } );
+				setBillingDetails( fieldValues );
+
+				if ( ! this.state.showDomainContactDetails ) {
+					setDomainDetails( fieldValues );
+					addGoogleAppsRegistrationData( fieldValues );
+				}
+			}
+
+			onComplete( error, messages );
+		};
+
+		wpcom.validateDomainContactInformation( fieldValues, this.getDomainNames(), validationHandler );
+	};
+
+	validateDomainDetails = ( fieldValues, onComplete ) => {
+		const validationHandler = ( error, data ) => {
+			const messages = ( data && data.messages ) || {};
+
+			if ( ! error ) {
+				setDomainDetails( fieldValues );
+				addGoogleAppsRegistrationData( fieldValues );
+			}
+
+			onComplete( error, messages );
+		};
+
+		if ( this.needsOnlyGoogleAppsDetails() ) {
+			wpcom.validateGoogleAppsContactInformation( fieldValues, validationHandler );
+			return;
+		}
+
+		wpcom.validateDomainContactInformation( fieldValues, this.getDomainNames(), validationHandler );
+	};
+
+	needsDomainDetails() {
+		const cart = this.props.cart;
+
+		if ( cart && cartItems.hasOnlyRenewalItems( cart ) ) {
+			return false;
+		}
+
+		return (
+			cart &&
+			( cartItems.hasDomainRegistration( cart ) ||
+				cartItems.hasGoogleApps( cart ) ||
+				cartItems.hasTransferProduct( cart ) )
+		);
+	}
+
+	needsOnlyGoogleAppsDetails() {
+		return (
+			cartItems.hasGoogleApps( this.props.cart ) &&
+			! cartItems.hasDomainRegistration( this.props.cart ) &&
+			! cartItems.hasTransferProduct( this.props.cart )
+		);
+	}
+
+	getDomainNames = () =>
+		map(
+			[
+				...cartItems.getDomainRegistrations( this.props.cart ),
+				...cartItems.getDomainTransfers( this.props.cart ),
+			],
+			'meta'
+		);
+
+	getTldsWithAdditionalForm() {
+		if ( ! config.isEnabled( 'domains/cctlds' ) ) {
+			// All we need to do to disable everything is not show the .FR form
+			return [];
+		}
+		return intersection( cartItems.getTlds( this.props.cart ), tldsWithAdditionalDetailsForms );
+	}
+
+	needsFax() {
+		return (
+			this.props.contactDetails.countryCode === 'NL' && cartItems.hasTld( this.props.cart, 'nl' )
+		);
+	}
+
+	renderSubmitButton() {
+		return (
+			<FormButton
+				className="checkout__domain-details-form-submit-button"
+				onClick={ this.handleSubmitButtonClick }
```

Comment:
Does `handleSubmitButtonClick` exist?

The forms differ in the way we display errors on the UI after a submission. Rather than preventing submission, we make a request and display a notification:

<img width="500" alt="screen shot 2018-11-02 at 2 09 48 pm" src="https://user-images.githubusercontent.com/6458278/47891933-64fde400-dea9-11e8-8bdd-3f0ce18ad8d4.png">

Also, this error is confusing since there are 2 x of every field.

Reply:
I removed some unnecessary methods that originally came from the `ContactDetailsForm` component.